tilavarauspalvelu/integrations/email/main.py

```diff
@@ -80,81 +137,64 @@ def send_application_received_email(
-        if not applications:
+    def send_application_section_cancelled(
+        application_section: ApplicationSection,
+        *,
+        language: Lang | None = None,
+    ) -> None:
+        """Sends an email that the whole application section was cancelled by the user"""
+        reservation: Reservation | None = application_section.actions.get_last_reservation()
+        if reservation is None:
             return
 
-        recipients_by_language = get_recipients_for_applications_by_language(applications)
-        if not recipients_by_language:
+        recipients = get_reservation_email_recipients(reservation=reservation)
+        if not recipients:
             SentryLogger.log_message(
-                "No recipients for application in allocation emails",
-                details={"applications": ",".join(str(application.pk) for application in applications)},
+                "No recipients for application section cancelled email",
+                details={"application_section": application_section.pk},
             )
             return
 
-        emails: list[EmailData] = []
-        email_type = EmailType.APPLICATION_IN_ALLOCATION
-
-        for language, recipients in recipients_by_language.items():
-            context = get_context_for_application_in_allocation(language=language)
-
-            emails.append(
-                EmailData(
-                    recipients=list(recipients),
-                    subject=context["title"],
-                    text_content=render_text(email_type=email_type, context=context),
-                    html_content=render_html(email_type=email_type, context=context),
-                    attachments=[],
-                )
-            )
+        if language is None:
+            language = get_reservation_email_language(reservation=reservation)
 
-        send_multiple_emails_in_batches_task.delay(emails=emails)
-        applications.update(in_allocation_notification_sent_date=local_datetime())
+        email_type = EmailType.APPLICATION_SECTION_CANCELLED
+        context = get_context_for_application_section_cancelled(application_section, language=language)
+        email = EmailData.build(recipients, context, email_type)
+        send_emails_in_batches_task.delay(email_data=email)
 
     @staticmethod
-    def send_application_handled_emails() -> None:
-        """Sends email to applicants when application round has its allocation results available."""
-        applications = Application.objects.should_send_handled_email().order_by("created_date")
-        if not applications:
-            SentryLogger.log_message("Zero applications require the 'application handled' email to be sent")
+    def send_staff_notification_application_section_cancelled(application_section: ApplicationSection) -> None:
+        """Sends an email to Staff that the whole application section was cancelled by the user"""
+        reservation: Reservation | None = application_section.actions.get_last_reservation()
```

Comment:
This misses the possibility that the application section contain multiple recurring reservation series at different (reservation) units, for which the recipients may be different (or the same, so need to deduce the unique recipients across them).

Reply:
Added a fix to get staff recipients for all reservation units options in the application section